Source Google Sheets: added unittest, updated to latest CDK, updated docs (#29491)
airbytehq/airbyte: docs/integrations/sources/google-sheets.md

# Google Sheets

> This page contains the setup guide and reference information for the Google Sheets source connector.

Author: updated google-sheets.inapp.md

:::info
The Google Sheets source connector pulls data from a single Google Sheets spreadsheet. To replicate multiple spreadsheets, set up multiple Google Sheets source connectors in your Airbyte instance.
:::

### Prerequisites
- Spreadsheet Link - The link to the Google spreadsheet you want to sync.

> - Authentication - Credentials for connecting to the Google Sheets API.

Author: added

   - To authenticate your Google account via Service Account Key Authentication, enter your [Google Cloud service account key](https://cloud.google.com/iam/docs/creating-managing-service-account-keys#creating_service_account_keys) in JSON format. Make sure the Service Account has the Project Viewer permission. If your spreadsheet is viewable by anyone with its link, no further action is needed. If not, [give your Service account access to your spreadsheet](https://youtu.be/GyomEw5a2NQ%22).
6. For **Spreadsheet Link**, enter the link to the Google spreadsheet. To get the link, go to the Google spreadsheet you want to sync, click **Share** in the top right corner, and click **Copy Link**.
7. For **Row Batch Size**, define the number of records you want the Google API to fetch at a time. The default value is 200.
8. For **Columns Name Conversion**, if set as true column names will be transformed. Examples: My_Name -> my_name, My Name -> my_name, MyName -> my_name, My123name -> my_123_name, привіт світ -> privit_svit.

### Output schema

Each sheet in the selected spreadsheet is synced as a separate stream. Each selected column in the sheet is synced as a string field.
    :::note

>     Sheet names and column headers must contain only alphanumeric characters or `_`, as specified in the** [**Airbyte Protocol**](../../understanding-airbyte/airbyte-protocol.md). For example, if your sheet or column header is named `the data`, rename it to `the_data`. This restriction does not apply to non-header cell values.

Author: thank you for pointing this out. source has convert for sheet columns(Convert Column Names to SQL-Compliant Format in config), but doesn't have convert for sheet names. the source works normally if the names of the sheets have not been converted to the standard. I’ve tested it.
From the source side we do not need to change the source code (we can convert sheet names too but it will be a breaking change). the Airbyte protocol says that destinations are responsible for cleaning these names to make them valid table and column names in their respective data stores. 
So I removed this note from the documentation.

- [Full Refresh - Overwrite](https://docs.airbyte.com/understanding-airbyte/connections/full-refresh-overwrite/)
- [Full Refresh - Append](https://docs.airbyte.com/understanding-airbyte/connections/full-refresh-append)

## Data type map

| Integration Type | Airbyte Type | Notes |
|:-----------------|:-------------|:------|
| any type         | `string`     |       |

## Performance consideration

The [Google API rate limit](https://developers.google.com/sheets/api/limits) is 100 requests per 100 seconds per user and 500 requests per 100 seconds per project. Airbyte batches requests to the API in order to efficiently pull data and respects these rate limits. We recommended not using the same service user for more than 3 instances of the Google Sheets source connector to ensure high transfer speeds.

## Troubleshooting

> - Data changes in spreadsheets. If your sheet becomes empty, deleted, etc., its synchronization will be stopped. You will see the corresponding extended logs about this problem. 

Author: updated description

> - Permissions. The connector cannot synchronize tables to which you do not have proper permissions.

Author: removed

- [Full Refresh - Overwrite](https://docs.airbyte.com/understanding-airbyte/connections/full-refresh-overwrite/)
- [Full Refresh - Append](https://docs.airbyte.com/understanding-airbyte/connections/full-refresh-append)

## Data type map

| Integration Type | Airbyte Type | Notes |
|:-----------------|:-------------|:------|
| any type         | `string`     |       |

## Performance consideration

The [Google API rate limit](https://developers.google.com/sheets/api/limits) is 100 requests per 100 seconds per user and 500 requests per 100 seconds per project. Airbyte batches requests to the API in order to efficiently pull data and respects these rate limits. We recommended not using the same service user for more than 3 instances of the Google Sheets source connector to ensure high transfer speeds.

## Troubleshooting

> - Data changes in spreadsheets. If your sheet is completely empty(no header rows) or deleted, Airbyte will stop attempting to sync it until it comes back. If this happens, the sync logs will contain a message saying the sheet has been skipped when syncing the full spreadsheet.

Reviewer: ```suggestion
- If your sheet is completely empty(no header rows) or deleted, Airbyte will not delete the table in the destination. If this happens, the sync logs will contain a message saying the sheet has been skipped when syncing the full spreadsheet.
```
